src/psyclone/tests/dynamo0p3_lma_test.py

```diff
@@ -0,0 +1,869 @@
+                                        "10.9_operator_first.f90"),
+                           api=TEST_API)
+    psy = PSyFactory(TEST_API, distributed_memory=True).create(invoke_info)
+    gen_code = str(psy.gen)
+    print(gen_code)
+    # Check that we use the field and not the operator to look-up the dofmap
+    assert "theta_proxy%vspace%get_whole_dofmap()" in gen_code
+    assert gen_code.count("get_whole_dofmap") == 1
+
+
+def test_operator_read_level1_halo():
+    ''' Check that we raise an error if a kernel attempts to read from an
+    operator beyond the level-1 halo '''
+    _, invoke_info = parse(os.path.join(BASE_PATH,
+                                        "10.7_operator_read.f90"),
+                           api=TEST_API)
+    psy = PSyFactory(TEST_API, distributed_memory=True).create(invoke_info)
+    schedule = psy.invokes.invoke_list[0].schedule
+    loop = schedule.children[0]
+    # Modify the loop bound so that we attempt to read from the L2 halo
+    # (of the operator)
+    loop.set_upper_bound("cell_halo", index=2)
+    # Attempt to generate the code
+    with pytest.raises(GenerationError) as excinfo:
+        _ = psy.gen
+    assert ("Kernel 'testkern_operator_code' reads from an operator and "
+            "therefore cannot be used for cells beyond the level 1 halo. "
+            "However the containing loop goes out to level 2" in str(excinfo))
+
+
+def test_operator_bc_kernel(tmpdir, f90, f90flags):
+    ''' Tests that a kernel with a particular name is recognised as a
+    kernel that applies boundary conditions to operators and that
+    appropriate code is added to support this. '''
+    _, invoke_info = parse(os.path.join(BASE_PATH,
+                                        "12.4_enforce_op_bc_kernel.f90"),
+                           api=TEST_API)
+    psy = PSyFactory(TEST_API, distributed_memory=True).create(invoke_info)
+    generated_code = str(psy.gen)
+    print(generated_code)
+    output1 = "INTEGER, pointer :: boundary_dofs_op_a(:,:) => null()"
+    assert output1 in generated_code
+    output2 = "boundary_dofs_op_a => op_a_proxy%fs_to%get_boundary_dofs()"
+    assert output2 in generated_code
+    output3 = (
+        "CALL enforce_operator_bc_code(cell, nlayers, op_a_proxy%ncell_3d, "
+        "op_a_proxy%local_stencil, ndf_any_space_1_op_a, "
+        "ndf_any_space_2_op_a, boundary_dofs_op_a)")
+    assert output3 in generated_code
+
+    # If compilation testing has been enabled
+    # (--compile --f90="<compiler_name>" flags to py.test)
+    assert code_compiles(TEST_API, psy, tmpdir, f90, f90flags)
+
+
+def test_operator_bc_kernel_fld_err(monkeypatch):
+    ''' Test that we reject the recognised operator boundary conditions
+    kernel if its argument is not an operator '''
+    _, invoke_info = parse(os.path.join(BASE_PATH,
+                                        "12.4_enforce_op_bc_kernel.f90"),
+                           api=TEST_API)
+    for dist_mem in [False, True]:
```

Comment:
fixture please and for all remaining in this file.

Reply:
Done.